File: COA-proj.html/script1.py
def fifo_cache(capacity, access_pattern):
    cache = []
    hits = 0
    misses = 0
    explanation = []

    for item in access_pattern:
        if item in cache:
            hits += 1
            explanation.append(f"Hit: {item} is already in the cache.")
        else:
            if len(cache) >= capacity:
                evicted = cache.pop(0)  # Remove the oldest item (FIFO)
                explanation.append(f"Evict: {evicted} removed to make space.")
            cache.append(item)
            misses += 1
            explanation.append(f"Miss: {item} added to the cache.")

    return {
        "hits": hits,
        "misses": misses,
        "final_cache": cache,
        "explanation": explanation
    }


def lru_cache(capacity, access_pattern):
    cache = []
    hits = 0
    misses = 0
    explanation = []

    for item in access_pattern:
        if item in cache:
            cache.remove(item)
            cache.insert(0, item)
            hits += 1
            explanation.append(f"Hit: {item} is moved to the front (most recently used).")
        else:
            if len(cache) >= capacity:
                evicted = cache.pop()
                explanation.append(f"Evict: {evicted} removed as it was least recently used.")
            cache.insert(0, item)
            misses += 1
            explanation.append(f"Miss: {item} added to the cache.")

    return {
        "hits": hits,
        "misses": misses,
        "final_cache": cache,
        "explanation": explanation
    }


def lfu_cache(capacity, access_pattern):
    cache = []
    frequency = {}
    hits = 0
    misses = 0
    explanation = []

    for item in access_pattern:
        if item in cache:
            frequency[item] += 1
            hits += 1
            explanation.append(f"Hit: {item} frequency increased.")
        else:
            if len(cache) >= capacity:
                # Find least frequently used item
                least_freq = min(frequency.values())
                candidates = [x for x in cache if frequency[x] == least_freq]
                to_evict = candidates[0]  # Tie-breaker: first in cache
                cache.remove(to_evict)
                del frequency[to_evict]
                explanation.append(f"Evict: {to_evict} removed due to low frequency.")
            cache.append(item)
            frequency[item] = 1
            misses += 1
            explanation.append(f"Miss: {item} added to the cache.")

    return {
        "hits": hits,
        "misses": misses,
        "final_cache": cache,
        "explanation": explanation
    }


def simulate_cache(capacity, policy, access_pattern):
    """
    Simulates cache replacement policies based on input.
    
    Args:
        capacity (int): Max size of the cache.
        policy (str): 'FIFO', 'LRU', 'LFU', or 'All'.
        access_pattern (list of str): Sequence of memory/page accesses.

    Returns:
        dict: Results with hits, misses, final cache, and explanation.
    """
    if not capacity or not access_pattern:
        return {"error": "Invalid input: Capacity and access pattern are required."}

    results = {}

    if policy == "All":
        results["FIFO"] = fifo_cache(capacity, access_pattern)
        results["LRU"] = lru_cache(capacity, access_pattern)
        results["LFU"] = lfu_cache(capacity, access_pattern)
    elif policy == "FIFO":
        results["FIFO"] = fifo_cache(capacity, access_pattern)
    elif policy == "LRU":
        results["LRU"] = lru_cache(capacity, access_pattern)
    elif policy == "LFU":
        results["LFU"] = lfu_cache(capacity, access_pattern)
    else:
        results["error"] = "Invalid policy selected."

    return results


# Interactive mode
if __name__ == "__main__":
    try:
        capacity = int(input("Enter cache capacity: "))
        policy = input("Choose policy (FIFO, LRU, LFU, All): ").strip().upper()
        raw_input = input("Enter access pattern (comma separated): ").strip()
        access_pattern = [item.strip() for item in raw_input.split(',') if item.strip()]

        result = simulate_cache(capacity, policy, access_pattern)

        for policy_name, data in result.items():
            print(f"\n--- {policy_name} Policy ---")
            print(f"Hits: {data['hits']}, Misses: {data['misses']}")
            print(f"Final Cache: {data['final_cache']}")
            print("Explanation:")
            for line in data["explanation"]:
                print("  " + line)

    except Exception as e:
        print("Error:", e)

File: script1.js
function simulateCache() {
    // Get input values from the form fields
    const capacity = Number(document.getElementById("capacity").value);
    const policy = document.getElementById("policy").value;
    const accessPattern = document.getElementById("input").value.split(',');

    // Validate input
    if (!capacity || !accessPattern.length) {
        alert("Please provide valid inputs.");
        return; // Stop execution if input is invalid
    }

    // Results for all policies
    let fifoResult = fifoCache(capacity, accessPattern);
    let lruResult = lruCache(capacity, accessPattern);
    let lfuResult = lfuCache(capacity, accessPattern);

    // Get the output div
    const output = document.getElementById("output");

    // Clear any previous output
    output.textContent = "";

    // If user selects "All", display results for all policies
    if (policy === "All") {
        // FIFO Policy Result
        output.textContent += `FIFO Policy Result:\nHits: ${fifoResult.hits}, Misses: ${fifoResult.misses}, Final Cache: ${JSON.stringify(fifoResult.final_cache)}\nExplanation:\n`;
        fifoResult.explanation.forEach(expl => output.textContent += expl + '\n');
        output.textContent += '\n';

        // LRU Policy Result
        output.textContent += `LRU Policy Result:\nHits: ${lruResult.hits}, Misses: ${lruResult.misses}, Final Cache: ${JSON.stringify(lruResult.final_cache)}\nExplanation:\n`;
        lruResult.explanation.forEach(expl => output.textContent += expl + '\n');
        output.textContent += '\n';

        // LFU Policy Result
        output.textContent += `LFU Policy Result:\nHits: ${lfuResult.hits}, Misses: ${lfuResult.misses}, Final Cache: ${JSON.stringify(lfuResult.final_cache)}\nExplanation:\n`;
        lfuResult.explanation.forEach(expl => output.textContent += expl + '\n');
    } 
    // If user selects FIFO, display only FIFO result
    else if (policy === "FIFO") {
        output.textContent += `FIFO Policy Result:\nHits: ${fifoResult.hits}, Misses: ${fifoResult.misses}, Final Cache: ${JSON.stringify(fifoResult.final_cache)}\nExplanation:\n`;
        fifoResult.explanation.forEach(expl => output.textContent += expl + '\n');
    } 
    // If user selects LRU, display only LRU result
    else if (policy === "LRU") {
        output.textContent += `LRU Policy Result:\nHits: ${lruResult.hits}, Misses: ${lruResult.misses}, Final Cache: ${JSON.stringify(lruResult.final_cache)}\nExplanation:\n`;
        lruResult.explanation.forEach(expl => output.textContent += expl + '\n');
    } 
    // If user selects LFU, display only LFU result
    else if (policy === "LFU") {
        output.textContent += `LFU Policy Result:\nHits: ${lfuResult.hits}, Misses: ${lfuResult.misses}, Final Cache: ${JSON.stringify(lfuResult.final_cache)}\nExplanation:\n`;
        lfuResult.explanation.forEach(expl => output.textContent += expl + '\n');
    }
}

function fifoCache(capacity, accessPattern) {
    let cache = [];
    let hits = 0;
    let misses = 0;
    let explanation = [];
    
    accessPattern.forEach(item => {
        if (cache.includes(item)) {
            hits++;
            explanation.push(`Hit: ${item} is already in the cache.`);
        } else {
            if (cache.length >= capacity) {
                let evicted = cache.shift(); // Remove the first (oldest) item
                explanation.push(`Evict: ${evicted} removed to make space.`);
            }
            cache.push(item); // Add new item to cache
            misses++;
            explanation.push(`Miss: ${item} added to the cache.`);
        }
    });

    return {
        hits,
        misses,
        final_cache: cache,
        explanation
    };
}

function lruCache(capacity, accessPattern) {
    let cache = [];
    let hits = 0;
    let misses = 0;
    let explanation = [];

    accessPattern.forEach(item => {
        if (cache.includes(item)) {
            cache.splice(cache.indexOf(item), 1); 
            cache.unshift(item);
            hits++;
            explanation.push(`Hit: ${item} is moved to the front (most recently used).`);
        } else {
            if (cache.length >= capacity) {
                let evicted = cache.pop(); // Remove the least recently used item
                explanation.push(`Evict: ${evicted} removed as it was least recently used.`);
            }
            cache.unshift(item); // Add new item to the front (most recently used)
            misses++;
            explanation.push(`Miss: ${item} added to the cache.`);
        }
    });

    return {
        hits,
        misses,
        final_cache: cache,
        explanation
    };
}

function lfuCache(capacity, accessPattern) {
    let cache = [];
    let frequency = {}; // Track the frequency of each item
    let hits = 0;
    let misses = 0;
    let explanation = [];

    accessPattern.forEach(item => {
        if (cache.includes(item)) {
            frequency[item]++; // Increment access count
            hits++;
            explanation.push(`Hit: ${item} frequency increased.`);
        } else {
            if (cache.length >= capacity) {
                let leastFrequentItem = Object.keys(frequency).reduce((a, b) => frequency[a] < frequency[b] ? a : b);
                cache.splice(cache.indexOf(leastFrequentItem), 1); // Remove least frequently used item
                delete frequency[leastFrequentItem];
                explanation.push(`Evict: ${leastFrequentItem} removed due to low frequency.`);
            }
            cache.push(item); // Add new item to the cache
            frequency[item] = 1; // Initialize its frequency
            misses++;
            explanation.push(`Miss: ${item} added to the cache.`);
        }
    });

    return {
        hits,
        misses,
        final_cache: cache,
        explanation
    };
}
